### ISSUE TITLE
xUnit1026 should support multiply-used member data

### ISSUE THREAD

**Issue by sharwell** [xunit/xunit#1556]
### Steps to reproduce

1. Define a property `Data` to provide member data which provides two arguments
2. Create two theories with `Data` as the test source
    * The first theory uses both parameters provided by `Data`
    * The second theory uses only one of the parameters

### Expected results

xUnit1026 is not reported. Even though the second theory does not use one of the parameters, the value itself must be provided by `Data` because it's used by another theory.

### Actual results

xUnit1026 is reported for the second theory. There is no way to fix the warning without breaking the test code.

**Comment by bradwilson** [comment]
This is not a very common scenario. Why not disable the rule on those rare occasions when it comes up?

**Comment by sharwell** [comment]
@bradwilson We're using a suppression for now, but it seems like a good case to leave open in case someone wants to implement it in the future. Sometimes false positives are inevitable, but in this case it seems straightforward to handle the majority of impacted cases.

**Comment by bradwilson** [comment]
The way analyzers work, this would be extremely difficult to do without a huge performance penalty. Anything which forces a scan is definitely discouraged.

**Comment by sharwell** [comment]
@bradwilson I'd be happy to trade you a fix if you can be specific about what you'd need made efficient in order to resolve this 😄

**Comment by bradwilson** [comment]
The efficiency you trade off here is how large a scope are you willing to check to disable the rule.

`[MemberData]` is not required to be on the same type as the test. Is it acceptable to only look at the other tests in the same class? What about all the tests in the assembly? What about all the tests in all the assemblies?

The open-ended nature of the requirements basically say that the only truly correct implementation causes you to scale all methods on all types in all assemblies, which you will never be able to do efficiently.

**Comment by sharwell** [comment]
Options which have less overhead include the following:

* If the `MemberData` references a member of another type, treat it as a data API and never report xUnit1026
2. 
* If `MemberData` references a member of the current type, treat it as an API for the current type, and only report xUnit1026 if no method in the current test class references the corresponding parameter.
    * If this is too expensive, treat the case the same as references to a member of another type

**Comment by bradwilson** [comment]
Well, @marcind is going to have to make the final call, but I would just keep it as-is.

**Comment by bradwilson** [comment]
Maybe an alternative would be to ignore unused parameters whose names consist entirely of underscores. Then you don't need to search anywhere.

**Comment by sharwell** [comment]
> Maybe an alternative would be to ignore unused parameters whose names consist entirely of underscores. 

I thought about doing this, but realized it would make it harder to understand what the unused data is, which makes it harder to understand during review why the data is unused by a particular test.

**Comment by marcind** [comment]
TLDR on my position: no to any analysis of `MemberData` or other sources applied to the theory as part of this rule, open to convention of ignoring parameters prefixed with `_`.

The purpose of the rule is to indicate to the test author that they may have a problem because they are not explicitly using all of the test data that's declared to be relevant to the test (i.e the theory method parameters). The problem could be that they are missing an assert or that the theory ends up being invoked multiple times for effectively equivalent data inputs - the extreme example being a theory that uses none of its parameters (i.e. what should be a `Fact` instead).

The current implementation does not care what (if any) data is passed into the theory, or the source of the data, only the data flow analysis within the function body. This keeps the implementation nice and fast and simple. I don't think that analyzing whether the theory is using a member-based data source that's used elsewhere (and how it's used in all those other places) is the right thing to do, especially when you consider that a particular theory method could be decorated with multiple data source attributes (including `InlineData` etc).

I must confess it's my personal opinion that having a big "spreadsheet" of data that you pass into lots of tests, each caring about some particular subsection thereof is an anti-pattern. I would most certainly create semantically meaningful projections of that data and feed those into the tests (so that, for example, changing the "column layout" in the master spreadsheet does not suddenly require fixing every single test, just the affected projection code).

Having said that, issues with this analyzer have been reported by others (e.g. https://twitter.com/natemcmaster/status/897536244346798081 and https://github.com/xunit/xunit/issues/1513). On the aspect of not firing for intentionally ignored parameters I've already [commented](https://github.com/xunit/xunit/issues/1513#issuecomment-336735772) that a convention of prefixing the param name with an `_` would be interesting (and would retain the param's meaning, as opposed to completely renaming it to be all underscores), though it does sound like using the VS tooling to put a pragma ignore around the affected parameter is also not that bad (at least in the latest updates of VS 2017, I will have to try it myself to see how bad it is).

**Comment by marcind** [comment]
Another vote for `_`: https://github.com/xunit/xunit/issues/1561

**Comment by bradwilson** [comment]
Closing this as fixed, as `^_\d*$` is now recognized for discard parameters: https://github.com/xunit/xunit.analyzers/blob/d62c883e68fdd5c12e397ba7760522ef95625b9e/src/xunit.analyzers/X1000/TheoryMethodShouldUseAllParameters.cs#L15
